zkevm-circuits/src/evm_circuit/util/common_gadget.rs

```diff
@@ -380,77 +380,63 @@ impl<F: Field> TransferToGadget<F> {
         cb: &mut EVMConstraintBuilder<F>,
         receiver_address: WordLoHi<Expression<F>>,
         receiver_exists: Expression<F>,
-        must_create: Expression<F>,
+        opcode_is_create: Expression<F>,
         value: Word32Cell<F>,
         mut reversion_info: Option<&mut ReversionInfo<F>>,
-        account_write: bool,
     ) -> Self {
         let value_is_zero = cb.is_zero_word(&value);
-        if account_write {
-            Self::create_account(
-                cb,
-                receiver_address.clone(),
-                receiver_exists.clone(),
-                must_create.clone(),
-                value_is_zero.expr(),
-                reversion_info.as_deref_mut(),
-            );
-        }
-        let receiver = cb.condition(not::expr(value_is_zero.expr()), |cb| {
-            UpdateBalanceGadget::construct(
-                cb,
-                receiver_address,
-                vec![value.clone()],
-                reversion_info,
-            )
-        });
-
-        Self {
-            receiver,
-            receiver_exists,
-            must_create,
-            value_is_zero,
-        }
-    }
 
-    pub(crate) fn create_account(
-        cb: &mut EVMConstraintBuilder<F>,
-        receiver_address: WordLoHi<Expression<F>>,
-        receiver_exists: Expression<F>,
-        must_create: Expression<F>,
-        value_is_zero: Expression<F>,
-        reversion_info: Option<&mut ReversionInfo<F>>,
-    ) {
+        // Create account
         cb.condition(
             and::expr([
                 not::expr(receiver_exists.expr()),
-                or::expr([not::expr(value_is_zero.expr()), must_create]),
+                or::expr([not::expr(value_is_zero.expr()), opcode_is_create.expr()]),
             ]),
             |cb| {
                 cb.account_write(
                     receiver_address.clone(),
                     AccountFieldTag::CodeHash,
                     cb.empty_code_hash(),
                     WordLoHi::zero(),
-                    reversion_info,
+                    reversion_info.as_deref_mut(),
                 );
             },
         );
+
+        let receiver = cb.condition(not::expr(value_is_zero.expr()), |cb| {
+            cb.increase_balance(receiver_address, value.clone(), reversion_info)
+        });
+
+        Self {
+            receiver,
+            receiver_exists,
+            opcode_is_create,
+            value_is_zero,
+        }
     }
 
     pub(crate) fn assign(
         &self,
         region: &mut CachedRegion<'_, '_, F>,
         offset: usize,
-        (receiver_balance, prev_receiver_balance): (U256, U256),
-        value: U256,
+        rws: &mut StepRws,
+        (receiver_exists, value, opcode_is_create): (bool, U256, bool),
     ) -> Result<(), Error> {
+        if !receiver_exists && (!value.is_zero() || opcode_is_create) {
+            let _receiver_code_hash = rws.next().account_codehash_pair();
```

Comment:
I think we don't need `_receiver_code_hash` here.

Reply:
We need a codehash write there. Otherwise, the next `rws.next()` would fail.
We can comment instead of using a dummy variable. Let me know which do you prefer.
```suggestion
            // receiver's code_hash
            rws.next().account_codehash_pair();
```

Reply:
Yeah, I know we need to call `rws.next()`. Using comment instead of a dummy variable looks good for me!